ic10emu/src/lib.rs

```diff
@@ -1120,32 +1183,56 @@ impl VM {
         }
     }
 
-    pub fn add_device_to_network(
+    pub fn set_device_connection(
         &self,
         id: u32,
-        network_id: u32,
         connection: usize,
+        target_net: Option<u32>,
     ) -> Result<bool, VMError> {
-        if let Some(network) = self.networks.get(&network_id) {
-            let Some(device) = self.devices.get(&id) else {
-                return Err(VMError::UnknownId(id));
+        let Some(device) = self.devices.get(&id) else {
+            return Err(VMError::UnknownId(id));
+        };
+        if connection >= device.borrow().connections.len() {
+            let conn_len = device.borrow().connections.len();
+            return Err(ICError::ConnectionIndexOutOfRange(connection, conn_len).into());
+        }
+
+        {
+            // scope this borrow
+            let connections = &device.borrow().connections;
+            let Connection::CableNetwork { net, .. } = & connections[connection] else {
+                return Err(ICError::NotACableConnection(connection).into());
             };
-            if connection >= device.borrow().connections.len() {
-                let conn_len = device.borrow().connections.len();
-                return Err(ICError::ConnectionIndexOutOfRange(connection, conn_len).into());
+            // remove from current network
+            if let Some(net) = net {
+                if let Some(network) = self.networks.get(net) {
+                    // if there is no other connection to this network
+                    if connections.clone().iter().enumerate().all(|(index, conn)| {
+                        if let Connection::CableNetwork { net: other_net, .. } = conn {
+                            !(other_net.is_some_and(|id| id == *net) && index != connection)
+                        } else {
+                            true
```

Comment:
The nested conditionals here are tricky.

optional idea: what if you instead `.filter()` for connections that _do_ match the network and then see if the `.count() == 1`. That'd cut down on the negations here which I think is why I struggle to read it.

Reply:
.. that would be a smart method